// fixedsticky/Gruntfile.js
module.exports = function(grunt) {
    'use strict';

    grunt.config.init({
        pkg: grunt.file.readJSON('package.json'),
        srcPath: 'src',
        buildPath: 'build',

        clean: ['<%= buildPath %>/*'],

        copy: {
            main: {
                files: [{
                    expand: true,
                    cwd: './',
                    src: ['package.json'],
                    dest: '<%= buildPath %>'
                }]
            }
        },

        concat: {
            main: {
                src: [
                    '<%= srcPath %>/sticky.js'
                ],
                dest: '<%= buildPath %>/sticky.debug.js'
            }
        },

        uglify: {
            main: {
                files: [{
                    expand: true,
                    cwd: '<%= buildPath %>',
                    src: '*.debug.js',
                    dest: '<%= buildPath %>',
                    ext: '.js'
                }]
            }
        },

        commonizor: {
            js: {
                files: {
                    '<%=buildPath%>/sticky.common.js':
                        ['<%=buildPath%>/sticky.js'],
                }
            }
        },

        watch: {
            main: {
                files: ['<%= srcPath %>/*.js'],
                tasks: ['concat', 'uglify']
            },
        },

    });

    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-copy');
    grunt.loadNpmTasks('grunt-contrib-concat');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-commonizor');

    grunt.registerTask('common', ['commonizor']);
    grunt.registerTask('default', ['clean', 'copy', 'concat', 'uglify', 'common']);
    grunt.registerTask('dev', ['watch']);
};
