// lazyimg/src/adapter.js
/*
    lib-img-adpter 
    Author: kongshi.wl@alibaba-inc.com 
    Date:   Dec,2015
*/
;

(function (win, lib) {
    require('appearjs');

    var adapter = {};
    var appearInstance;
    var runtimeFlags = {};

    var config = {
        'dataSrc': 'img-src', //指定图片地址的attribute名, 兼做lazy-class的作用
        'lazyHeight': 0, //以此高度提前触发懒加载
        'lazyWidth': 0 //以此宽度提前触发懒加载
    };


    function extendStrict(main, sub) {
        var ret = {};
        for (var k in main) {
            if (main.hasOwnProperty(k)) {
                ret[k] = sub.hasOwnProperty(k) ? sub[k] : main[k];
            }
        }
        return ret;
    }

    function applySrc(item, processedSrc) {
        if (!processedSrc) {
            return;
        }
        if (item.nodeName.toUpperCase() == 'IMG') {
            item.setAttribute('src', processedSrc);
        } else {
            item.style.backgroundImage = 'url("' + processedSrc + '")';
        }
    }

    function init() {
        appearInstance = lib.appear.init({
            cls: 'imgtmp', //可选，需要遍历的元素
            once: true, //可选，是否只触发一次
            x: config.lazyWidth, //可选，容器右边距离x以内的元素加载，默认为0
            y: config.lazyHeight, //可选，容器底部距离y以内的元素加载，默认为0
            onAppear: function (evt) {
                var item = this;
                applySrc(item, item.getAttribute('i-lazy-src'));
                item.removeAttribute('i-lazy-src');
            }
        });
    }


    adapter.logConfig = function () {
        console.log('lib-img Config\n', config);
    }


    adapter.fire = function () {

        if (!appearInstance) {
            init();
        }

        var label = 'i_' + Date.now() % 100000;
        var domList = document.querySelectorAll('[' + config.dataSrc + ']');

        [].forEach.call(domList, function (item) {
            if (item.dataset.lazy == 'false' && item.dataset.lazy != 'true') {
                applySrc(item, processSrc(item, item.getAttribute(config.dataSrc)));
            } else {
                item.classList.add(label);
                item.setAttribute('i-lazy-src', item.getAttribute(config.dataSrc));
            }
            item.removeAttribute(config.dataSrc);
        });

        appearInstance.bind('.' + label);
        appearInstance.fire();
    }



    adapter.defaultSrc = 'data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgYGBgAAAABQABh6FO1AAAAABJRU5ErkJggg==';



    lib.img = adapter;
    
    module.exports = adapter;

})(window, window['lib'] || (window['lib'] = {}));
